notifier/webhook/deliverer.go

```diff
@@ -0,0 +1,144 @@
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/google/uuid"
+	clairerror "github.com/quay/clair/v4/clair-error"
+	"github.com/quay/clair/v4/notifier"
+	"github.com/quay/clair/v4/notifier/keymanager"
+	"github.com/rs/zerolog"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+type Deliverer struct {
+	conf Config
+	// a client to use for POSTing webhooks
+	c    *http.Client
+	kmgr *keymanager.Manager
+}
+
+// New returns a new webhook Deliverer
+func New(conf Config, client *http.Client, keymanager *keymanager.Manager) (*Deliverer, error) {
+	var c Config
+	var err error
+	if c, err = conf.Validate(); err != nil {
+		return nil, err
+	}
+	if client == nil {
+		client = http.DefaultClient
+	}
+	return &Deliverer{
+		conf: c,
+		c:    client,
+		kmgr: keymanager,
+	}, nil
+}
+
+func (d *Deliverer) Name() string {
+	return "webhook"
+}
+
+// sign will use the provided private key to sign and attach a jwt to the provided
+// request.
+//
+// a kid header specifying the private key which signed the jwt is attached as well.
+func (d *Deliverer) sign(ctx context.Context, req *http.Request, kp keymanager.KeyPair) error {
+	signer, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.PS512, Key: kp.Private}, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		return fmt.Errorf("failed to create jwt signer: %v", err)
+	}
+	now := jwt.NumericDate(time.Now().Unix())
+	expire := jwt.NumericDate(time.Now().Add(1 * time.Hour).Unix())
+	cl := jwt.Claims{
+		Subject:  "clair-notifier",
+		Issuer:   "clair-notifier",
```

Comment:
I think this clashes with the hardcoded path in the notifier. This would result in requests for `/services/clair-notifier/keys/{id}`

Reply:
I'm not seeing how issuer field clashes with anything else in the app? Isn't the only thing concerned with this value is the receiver of the JWT. Does "issuer" get used for anything other then the receiving code's verification logic.

Reply:
because the service name is hardcoded to just "notifier" in the path, but this would cause the verifier to look up "clair-notifier"